lib/luacrypto_aead.c

```diff
@@ -0,0 +1,258 @@
+*/
+static int luacrypto_aead_setkey(lua_State *L) {
+	struct crypto_aead *tfm = luacrypto_aead_check(L, 1);
+	size_t keylen;
+	const char *key = luaL_checklstring(L, 2, &keylen);
+	int ret = crypto_aead_setkey(tfm, key, keylen);
+	if (ret < 0)
+		luaL_error(L, "Failed to set AEAD key (err %d)", ret);
+	return 0;
+}
+
+/***
+* Sets the authentication tag size for the AEAD transform.
+* @function crypto_aead:setauthsize
+* @tparam integer tagsize The desired authentication tag size in bytes.
+* @raise Error if setting the authsize fails (e.g., unsupported size).
+*/
+static int luacrypto_aead_setauthsize(lua_State *L) {
+	struct crypto_aead *tfm = luacrypto_aead_check(L, 1);
+	unsigned int tagsize = luaL_checkinteger(L, 2);
+	int ret = crypto_aead_setauthsize(tfm, tagsize);
+	if (ret < 0)
+		luaL_error(L, "Failed to set AEAD authsize (err %d)", ret);
+	return 0;
+}
+
+/***
+* Gets the required initialization vector (IV) size for the AEAD transform.
+* @function crypto_aead:ivsize
+* @treturn integer The IV size in bytes.
+*/
+static int luacrypto_aead_ivsize(lua_State *L) {
+	struct crypto_aead *tfm = luacrypto_aead_check(L, 1);
+	lua_pushinteger(L, crypto_aead_ivsize(tfm));
+	return 1;
+}
+
+/***
+* Gets the current authentication tag size for the AEAD transform.
+* This is the value set by `setauthsize` or the algorithm's default.
+* @function crypto_aead:authsize
+* @treturn integer The authentication tag size in bytes.
+*/
+static int luacrypto_aead_authsize(lua_State *L) {
+	struct crypto_aead *tfm = luacrypto_aead_check(L, 1);
+	lua_pushinteger(L, crypto_aead_authsize(tfm));
+	return 1;
+}
+
+/* Auxiliary functions to define luacrypto_aead_encrypt and luacrypto_aead_decrypt */
+
+static inline size_t luacrypto_aead_get_encrypt_buffer_needed(lua_State *L, size_t combined_len, unsigned int authsize_val, size_t crypt_len) {
+	return combined_len + authsize_val;
+}
+
+static inline size_t luacrypto_aead_get_decrypt_buffer_needed(lua_State *L, size_t combined_len, unsigned int authsize_val, size_t crypt_len) {
+	luaL_argcheck(L, crypt_len >= authsize_val, 3, "input data (ciphertext+tag) too short for tag");
+	return combined_len;
+}
+
+static inline int luacrypto_aead_common_crypt(lua_State *L, int (*crypt_func)(struct aead_request *req), size_t (*get_buffer_needed_func)(lua_State *L, size_t combined_len, unsigned int authsize_val, size_t crypt_len), int result_adjust_factor)
+{
+	struct crypto_aead *tfm = luacrypto_aead_check(L, 1);
+	size_t iv_len, combined_len;
+	const char *iv = luaL_checklstring(L, 2, &iv_len);
+	const char *combined = luaL_checklstring(L, 3, &combined_len);
+	lua_Integer aad_l = luaL_checkinteger(L, 4);
+
+	unsigned int expected_ivlen = crypto_aead_ivsize(tfm);
+	luaL_argcheck(L, iv_len == expected_ivlen, 2, "incorrect IV length");
+	u8 *iv_data = lunatik_checkalloc(L, iv_len);
+	memcpy(iv_data, iv, iv_len);
+
+	luaL_argcheck(L, aad_l >= 0 && (size_t)aad_l <= combined_len, 4, "AAD length out of bounds");
```

Comment:
please notice, that there are a bunch of potential error cases like this one where you are going to leak `iv_data`.. I've pointed a very similar situation in your previous PR on the usage of luaL_Buffer.. I think you should be more careful about this.. can you also take the opportunity and review the already merged crypto libs to make sure we don't have similar conditions there? thanks!